File: ChordVoicingsCreatorJava/src/main/java/com/campoy/chord/voicing/creator/model/musictheory/ChordConstructor.java
package com.campoy.chord.voicing.creator.model.musictheory;

import java.util.function.Function;

public class ChordConstructor {
    private final String name;
    private final Function<Note, Chord> chordConstructingfunction;
    
    public ChordConstructor(String name, Function<Note, Chord> chordConstructingfunction) {
        this.name = name;
        this.chordConstructingfunction = chordConstructingfunction;
    }
    
    public Chord constructChord(Note note) {
        return chordConstructingfunction.apply(note);
    }
    
    public String getName() {
        return name;
    }
    
    @Override
    public boolean equals(Object other) {
        
        if( !( other instanceof ChordConstructor) ) {
            return false;
        }
        
        ChordConstructor otherChordConstructor = (ChordConstructor)other;
        
        Chord thisChord = this.constructChord(Note.A);
        Chord otherChord = otherChordConstructor.constructChord(Note.A);
        
        if(thisChord == null) {
        	return otherChord == null;
        }
        
        return (thisChord.equals(otherChord));
        
    }
    
    
    
}

File: PancakeMusicProjectsTS/src/model/musictheory/ChordConstructor.ts
import { EqualsTrait } from '../general/EqualsTrait.js';
import { Chord } from './Chord.js';
import { Note } from './Note.js';

export class ChordConstructor implements EqualsTrait{
    private readonly name: string;
    private readonly chordConstructingfunction: (note: Note) => Chord;
  
    constructor(name: string, chordConstructingfunction: (note: Note) => Chord) {
      this.name = name;
      this.chordConstructingfunction = chordConstructingfunction;
    }
  
    constructChord(note: Note): Chord {
      return this.chordConstructingfunction(note);
    }
  
    getName(): string {
      return this.name;
    }
  
    equals(other: any): boolean {
      if (!(other instanceof ChordConstructor)) {
        return false;
      }
  
      const thisChord = this.constructChord(Note.A);
      const otherChord = other.constructChord(Note.A);
  
      if (thisChord === null) {
        return otherChord === null;
      }
  
      return thisChord.equals(otherChord);
    }
  }